File: Java/src/_1239_Maximum_Length_of_a_Concatenated_String_with_Unique_Characters/Solution.java
package _1239_Maximum_Length_of_a_Concatenated_String_with_Unique_Characters;

import _0000_study._common.Build;

import java.util.List;

/**
 * Date: 2021/6/19 9:51
 * 执行用时：2 ms, 在所有 Java 提交中击败了97.75%的用户
 * 内存消耗：35.9 MB, 在所有 Java 提交中击败了87.08%的用户
 */
public class Solution {

    int[] state;
    boolean[] st;  // st[i]=true说明arr[i]自身有重复元素
    int res = 0;

    public int maxLength(List<String> arr) {
        int n = arr.size();
        state = new int[n];
        st = new boolean[n];
        for (int i = 0; i < n; i++) {
            for (char c : arr.get(i).toCharArray()) {
                int bit = 1 << (c - 'a');
                if ((state[i] & bit) != 0) {
                    st[i] = true;
                    break;
                }
                state[i] |= bit;
            }
        }
        dfs(arr, 0, 0, 0);
        return res;
    }

    void dfs(List<String> arr, int u, int cnt, int t) {
        if (u == arr.size()) {
            res = Math.max(res, cnt);
            return;
        }
        dfs(arr, u + 1, cnt, t);  // 不选arr[u]
        if ((t & state[u]) == 0 && !st[u]) {
            dfs(arr, u + 1, cnt + arr.get(u).length(), t | state[u]);  // 选arr[u]
        }
    }

    public static void main(String[] args) {

        List<String> arr = Build.build(new String[]{"un", "iq", "ue"});
        System.out.println((new Solution()).maxLength(arr));  // 4
    }
}

File: Python/_1239_Maximum_Length_of_a_Concatenated_String_with_Unique_Characters/Solution.py
# coding=utf-8
# Date: 2021/6/19 9:59
from typing import List


# 执行用时：88 ms, 在所有 Python3 提交中击败了84.72%的用户
# 内存消耗：15.1 MB, 在所有 Python3 提交中击败了59.80%的用户
class Solution:
    def __init__(self):
        self.state = []
        self.st = []
        self.res = 0

    def maxLength(self, arr: List[str]) -> int:
        n = len(arr)
        self.state = [0 for _ in range(n)]
        self.st = [False for _ in range(n)]
        for i in range(n):
            for c in arr[i]:
                bit = 1 << (ord(c) - ord('a'))
                if (self.state[i] & bit) != 0:
                    self.st[i] = True
                    break
                self.state[i] |= bit
        self.dfs(arr, 0, 0, 0)
        return self.res

    def dfs(self, arr, u, cnt, t):
        if u == len(arr):
            self.res = max(self.res, cnt)
            return
        self.dfs(arr, u + 1, cnt, t)
        if (t & self.state[u]) == 0 and not self.st[u]:
            self.dfs(arr, u + 1, cnt + len(arr[u]), t | self.state[u])


if __name__ == "__main__":
    print(Solution().maxLength(["un", "iq", "ue"]))  # 4
